lib/Backend/FlowGraph.cpp

```diff
@@ -4431,6 +4431,311 @@ BasicBlock::CleanUpValueMaps()
+    // If we find a Ld_A, we find if the src1 already has a value in the local valueTable, if not we use the value from the pred's valueTable merge
+    // For other instrs, we assign null value
+    GlobHashTable * localSymToValueMap = GlobHashTable::New(globOpt->alloc, 8);
+
+    FOREACH_INSTR_IN_BLOCK(instr, this)
+    {
+        if (OpCodeAttr::HasDeadFallThrough(instr->m_opcode))
+        {
+            return nullptr;
+        }
+        if (instr->GetDst())
+        {
+            if (instr->GetDst()->GetStackSym())
+            {
+                if (instr->m_opcode == Js::OpCode::Ld_A)
+                {
+                    Value **localValue = localSymToValueMap->FindOrInsertNew(instr->GetDst()->GetSym());
+                    if (instr->GetSrc1()->GetSym())
+                    {
+                        Value **localSrc1Value = localSymToValueMap->Get(instr->GetSrc1()->GetSym());
+                        *localValue = localSrc1Value == nullptr ? this->globOptData.FindValue(instr->GetSrc1()->GetSym()) : *localSrc1Value;
+                    }
+                }
+                else
+                {
+                    // complex instr, can't track value, insert nullptr
+                    Value **localValue = localSymToValueMap->FindOrInsertNew(instr->GetDst()->GetSym());
+                    *localValue = nullptr;
+                }
+            }
+        }
+    } NEXT_INSTR_IN_BLOCK;
+
+    IR::Instr * instr = this->GetLastInstr();
+    IR::LabelInstr * currentLabel = nullptr;
+    /* We start from the current instruction and go on scanning for legality, as long as it is legal to skip an instruction, skip.
+     * When we see an unconditional branch, start scanning from the branchTarget
+     * When we see a conditional branch, check if we can prove the branch target, if we can, adjust the flowgraph, and continue in the direction of the proven target
+     * We stop, when we no longer can skip instructions, either due to legality check or a non provable conditional branch
+    */
+    while (instr)
+    {
+        if (!instr->IsBranchInstr() && !instr->IsLabelInstr() && !IsLegalOpcodeForPathDepBrFold(instr))
+        {
+            return lastBranchTarget;
+        }
+        if (OpCodeAttr::HasDeadFallThrough(instr->m_opcode)) // BailOnNoProfile etc
+        {
+            return lastBranchTarget;
+        }
+        if (instr->GetDst())
+        {
+            if (!instr->GetDst()->IsRegOpnd()) // complex dstOpnd, stop.
+            {
+                return lastBranchTarget;
+            }
+            IR::RegOpnd *dst = instr->GetDst()->AsRegOpnd();
+            if (currentLabel)
+            {
+                BasicBlock * currentBlock = currentLabel->GetBasicBlock();
+                // If the dstOpnd is defined and used in the same block, it is legal, but if the successor blocks use it, stop.
+                if (currentBlock->succUses && currentBlock->succUses->Test(dst->GetStackSym()->m_id))
```

Comment:
We need isDead outside the currentBlock, isDead on the Opnd is too strict, so I have added a new bit vector which is upwardExposedUses of the successor blocks